경북대 FE 장세현 3단계 - 단체 주문 기능 구현하기

SeHIgh opened the pull request:
안녕하십니까 경북대학교 프론트엔드 과정 수강중인 장세현입니다.
3단계 과제 - 단체 주문 기능 구현하기 내용을 PR 요청 드립니다!

이번 과제 진행하기 앞서서
기존의 폼과 유효성 검증 코드의 복잡성과 구현하는 기능이 증가하면서 프로젝트 구조 관리가 힘든 점이있어서
FSD 아키텍쳐(https://dev.to/m_midas/feature-sliced-design-the-best-frontend-architecture-4noj) 를 정하고, 그에 맞게 프로젝트 구조를 먼저 개선하고 과제를 진행하였습니다!
(FSD 아키텍쳐는 이전에 강의 때 언급을 해주셔서 이용해보았습니다!)

최대한 아래 계층 구조로 관리하고자 노력하였고, 상위 계층만 하위 계층을 참조하도록 노력하였습니다.

```
app > pages > widgets > features > entities > shared
```

이전 코드를 다 분해하고 재작성하기에는 너무 힘들 것 같아
기존 코드의 디렉토리 위치만 수정하는 형태로 아키텍쳐를 최대한 지키도록 하였습니다.
(기존 코드를 옮길때 LLM의 힘을 조금 빌렸습니다..!)

### 개선된 프로젝트 구조 트리
```
react-gift-order/
└── src/
    ├── app/                        # 앱 시작점 (라우팅, HOC)
    │   ├── App.tsx                 # 메인 앱 컴포넌트
    │   ├── Router.tsx              # 페이지 라우팅 설정
    │   ├── main.tsx                # 앱 엔트리포인트
    │   └── hoc/
    │       └── withAuth.tsx        # 인증 고차 컴포넌트
    ├── entities/                   # 비즈니스 도메인 (데이터 모델, 스키마)
    │   ├── order/                  # 주문 관련 엔티티
    │   │   ├── orderFormSchema.ts      # 주문 폼 검증 스키마 (이번 과제 주요 부분!)
    │   │   └── orderCardMock.ts        # 주문 카드 목 데이터
    │   ├── product/                # 상품 관련 엔티티
    │   │   ├── categoryMock.ts         # 카테고리 목 데이터
    │   │   └── productListMock.ts      # 상품 목록 목 데이터
    │   └── user/                   # 유저 관련 엔티티
    │       └── loginSchema.ts          # 로그인 검증 스키마
    ├── features/                   # 핵심 기능 (사용자 시나리오별)
    │   ├── order/                  # 주문 관련 기능 (이번 과제 주요 부분!)
    │   │   ├── OrderCardSection.tsx        # 주문 카드 섹션
    │   │   ├── ReceiverSection.tsx         # 수신자 관리 섹션
    │   │   ├── SenderSection.tsx           # 발신자 정보 섹션
    │   │   └── Receiver/                   # 수신자 관련 하위 기능
    │   │       ├── ReceiverAddModal.tsx        # 수신자 추가 모달
    │   │       └── ReceiverTable.tsx           # 수신자 테이블
    │   └── product/                # 상품 관련 기능
    │       └── trend/              # 트렌드 상품 기능
    │           ├── Trend.tsx               # 트렌드 메인 컴포넌트
    │           ├── TrendFilter.tsx         # 트렌드 필터
    │           └── ProductItem.tsx         # 상품 아이템
    ├── pages/                      # 페이지 컴포넌트 (라우트별)
    │   ├── Home.tsx                # 홈 페이지
    │   ├── Login.tsx               # 로그인 페이지
    │   ├── Order.tsx               # 주문 페이지
    │   └── MyPage.tsx              # 마이 페이지
    ├── shared/                     # 공통 재사용 코드
    │   ├── contexts/               # React Context (전역 상태)
    │   │   ├── auth/AuthContext.tsx            # 인증 상태 관리
    │   │   └── order/OrderFormContext.tsx      # 주문 폼 상태 관리
    │   ├── hooks/                  # 커스텀 훅
    │   │   └── useInput.ts             # 입력 폼 훅
    │   ├── lib/                    # 유틸리티 함수
    │   │   ├── cookieUtils.ts          # 쿠키 관리
    │   │   ├── validateRules.ts        # 검증 규칙
    │   │   └── validateValue.ts        # 검증 함수
    │   ├── styles/                 # 스타일 시스템
    │   │   ├── theme.ts            # 테마 설정 (색상, 간격 등)
    │   │   ├── typography.ts       # 타이포그래피
    │   │   └── GlobalStyles.tsx    # 전역 스타일
    │   └── ui/                     # 공통 UI 컴포넌트
    │       ├── Button.tsx          # 버튼 컴포넌트
    │       ├── Modal.tsx           # 모달 컴포넌트
    │       ├── Table.tsx           # 테이블 컴포넌트
    │       └── Typography.tsx      # 텍스트 컴포넌트
    └── widgets/                    # 페이지 내 조합 UI 블록
        ├── header/Nav.tsx          # 네비게이션 헤더
        ├── banner/Banner.tsx       # 배너 위젯
        ├── category/Category.tsx   # 카테고리 위젯
        └── friends/Friends.tsx     # 친구 목록 위젯
```

 ---

### 작업 내용

- `entities/order/` 과 `features/order/`, `shared/ui/Modal.tsx` 경로가 주요 작업 부분이라 리뷰해주실 때 참고 부탁드립니다!

- React Hook Form & Zod 를 사용하여 폼과 유효성 검증 개선


- FSD 아키텍쳐를 적용하여 코드 관리 방법 개선하기
- 주문하기 페이지의 폼을 React Hook Form를 사용해서 리팩터링하기. (form 데이터를 useState로 다루는 경우는 없도록 하기.)

- 받는 사람 기능 고도화
  - 참고 URL을 살펴보고 받는 사람 기능의 UI를 구현
    - 모달 컴포넌트 만들기
    - 받는 사람 섹션 형태 수정하기
      - 추가 버튼 및 받는 사람 목록 나타내기
  - 받는 사람을 최대 10명까지 등록 가능하게 하기
  - 10명의 정보가 정확하게 입력되지 않으면 정보가 반영되지 않도록 하기
  - 받는 사람의 전화번호가 중복되면 안되게 하기
  - 최소 선택 수량은 1개 이상이도록 하기
  - 전화번호는 01012341234 형태로만 받을 수 있도록 하기
      
- 주문하기 페이지 고도화
  - 참고 URL을 살펴보고 변경된 기능을 구현하기
  - 유효성 검사를 Zod 라이브러리를 사용해서 처리하기

- 본인만의 기준으로 일관된 코드 작성하기
- 기능 단위로 나누어 커밋하기
- 재사용 가능성이 높은 컴포넌트, Hooks를 재사용성을 고려하여 구현하기

SeHIgh committed:
docs(README): 주문하기 페이지의 폼 리팩터링 및 받는 사람 기능 고도화 요구사항 추가

SeHIgh committed:
refactor(FSD): Feature-Sliced Design 아키텍처 적용 및 인증 시스템 개선

- FSD 계층 구조로 전면 리팩터링 진행 (app > pages > widgets > features > entities > shared)
- AuthContext에 isInitialized 상태 추가하여 URL 직접 접속 시 Race Condition 해결
- FSD 표준에 맞게 shared/utils → shared/lib 리네이밍
- 각 슬라이스별 Public API(index.ts) 추가로 계층 캡슐화 강화
- 모든 import 경로를 새로운 FSD 구조에 맞게 업데이트
- FSD 계층 의존성 규칙 준수 (상위 계층만 하위 계층 참조)

주요 개선사항:
- 로그인 상태 초기화 완료까지 대기하도록 withAuth HOC 개선
- 계층별 명확한 책임 분리로 코드 유지보수성 향상

SeHIgh committed:
refactor(order): 주문하기 페이지의 폼 리팩터링 (react-hook-form & zod 사용)

- react-hook-form 와 zod 를 사용하여 폼 상태 관리 및 유효성 검사를 개선
- 기존 상태 관리 로직 제거 및 Controller 컴포넌트로 대체
- react-hook-form, zod, @hookform/resolvers 의존성 추가

SeHIgh committed:
refactor(login): 로그인 페이지 폼 리팩터링 (react-hook-form & zod 사용)

+ 주문하기 폼 스키마 위치 변경 (features/order -> entities/order)

SeHIgh committed:
모달 생성 임시 커밋

SeHIgh committed:
feat(modal): 재사용 가능한 모달 구현

- 오버레이, 모달 구현
- 기본적으로 닫기버튼이 활성화 되어있고, false로 끌 수 있음
- 사용자가 원한다면 prop으로 헤더, 푸터 내용을 추가할 수 있음

SeHIgh committed:
feat(order): 주문 수신자 관리 및 폼 스키마 개선

- 주문 수신자 추가/수정 모달 및 테이블 컴포넌트 추가
- 주문 폼 스키마 및 타입 보완
- 주문 페이지 UI 및 기능 개선

SeHIgh committed:
refactor: 인코딩 유틸 및 App 구조 리팩토링

- encodingUtils 함수 개선
- App 컴포넌트 구조 정리

SeHIgh committed:
feat(ui): 공통 Modal, Table 컴포넌트 및 스타일 개선

- Table 컴포넌트 신규 추가 및 Modal 개선
- 공통 UI index 및 타이포그래피 스타일 보완
- Friends, Nav 위젯 UI 개선

SeHIgh committed:
docs: README 최신화

SeHIgh committed:
feat(auth): 인증 HOC 및 로그인/마이페이지 개선

- 인증 HOC 로직 개선
- 로그인 페이지 UI 및 검증 로직 보완
- 마이페이지 접근 제어 및 UI 개선
- 불필요한 AuthContext 파일 정리

SeHIgh committed:
feat(auth): AuthContext 파일 위치 수정

- FSD 아키텍쳐 준수를 위해 AuthContext 파일 위치 이동

terajh commented on src/features/order/Receiver/ReceiverAddModal.tsx:
```diff
@@ -0,0 +1,378 @@
+
+    if (hasDuplicate) {
+      return // 중복이 있으면 제출하지 않음
+    }
```
👍 Map 등을 활용해 깔끔한 형태로 중복된 전화번호를 처리해주셨군요ㅎㅎ
같은 로직을 zod 을 활용하는 형태로 한번 고민해보시는것 추천드립니다! zod 의 refine 을 활용하면 ReceiverAddModal 내에 있는 중복 체크에 대한 책임을 잘 격리시켜줄 수 있을 것 같아요

terajh commented on src/entities/order/orderFormSchema.ts:
```diff
@@ -0,0 +1,45 @@
+import { z } from 'zod'
```
[소소의견] entities layer 안에 order, product slice 를 구분지어 분리해주신것 같은데, 하나의 order slice 폴더 내에 다른 성격의 파일들이 같은 위치에 있는것 같아요! 성격별로 segment 를 두어 구분지어주어도 괜찮을 것 같다는 생각이 들어 코멘트 남겨봅니다! order slices를 기준으로는 data, schema 와 같은 segment 폴더를 두어 좀더 정리해볼 수도 있을 것 같네요ㅎㅎ (현구조도 좋습니다)

세현님은 어떻게 생각하시나요? 😃

SeHIgh replied:
지금 보니 멘토님 말씀대로 Mock 데이터와 Schema는 segment 폴더를 통해 나누어 관리하는 쪽이 더 좋을 것 같습니다.
이미 뭔가 좀 많이 나누어 둔 상태인데 더 나누어 관리하려니 파일 구조가 더 복잡해 지지 않을까 염려스러웠지만, 이번에 FSD 아키텍쳐를 사용해보면서 폴더 구조가 늘어나더라도 뭔가 확실하게 레이어를 나누고 책임 분리를 통하여 나중에 유지보수하기 쉽도록 하는게 좋은 방식이라는 생각이 들었습니다.

terajh commented on src/entities/order/orderFormSchema.ts:
```diff
@@ -0,0 +1,45 @@
+
+// * 주문하기 폼 스키마 정의 (Zod 사용)
+export const orderFormSchema = z.object({
+  // * 카드 메시지 (message 규칙 사용)
```
👍 적절하게 주석을 사용해 주시곤하는데, 코드 읽는 시간을 줄일수 있는 좋은 습관으로 보입니다! 다만 과도한 주석의 경우 오히려 코드 리딩에 악영향을 끼칠 수 있다보니 그점만 염두해주셔도 좋을 것 같습니다! 👍

SeHIgh replied:
넵! 저도 점점 기능을 구현하다 보니 나중에 봤을 때, 이해하기 쉽도록 제가 구현한 내용을 주석으로 조금 상세히, 많이 남겨두고 있는데 하나하나 이해는 쉬울지몰라도 전체적으로는 조금 난잡해 보이기도 하는 것 같습니다.
코드 자체에 이렇게 주석을 많이 남기기보다 진짜 필요한 주석만 남기고
세부 내용은 따로 MD 파일이나 text 형태로 남겨두는 것이 좋을 것 같다 생각이 듭니다!

terajh reviewed (changes requested):
안녕하세요 세현님! 3단계 미션 고생 많으셨습니다 👍🏻

React Hook Form 활용과 받는 사람 기능 고도화 등 3단계 미션의 요구사항들을 전반적으로 잘 구현해주셨네요. 👍 추가로 FSD 아키텍처에 대해 고민하고 실제 적용까지 해보시면서 다양한 인사이트를 얻으신 것 같아 인상 깊었습니다.

FSD 아키텍처는 비교적 최근에 주목받고 있는 구조로, 저도 개인적으로는 트렌디한 아키텍처라고 생각해요. 다만 그만큼 아직 완전히 정형화된 정답이 있는 구조는 아니라고 생각해요, 팀마다 또는 개인마다 각자의 컨벤션에 맞춰 커스터마이징해서 사용하는 경우가 많다고 알고 있습니다.

때문에 “FSD는 무조건 이렇게!”보다는, 내가 어떤 기준으로 파일을 나누고 어떤 의도를 가지고 구조를 설계했는지 그 과정을 고민하고 경험해보는 것이 더 큰 가치라고 생각합니다.
이번 미션이 그런 의미에서 좋은 학습 기회가 되었기를 바랍니다! 💪

참고로 몇가지 고민해보면 좋을 코멘트들을 작성해보았는데요. 확인 후 다시 리뷰요청 부탁드립니다. 고생하셨습니다~~

SeHIgh committed:
refactor(order): Entities 계층의 데이터 및 스키마 분할 리팩토링

SeHIgh committed:
refactor(order):  중복 전화번호 검증 로직을 zod 스키마에 위임하여 통합관리

- 주문 수신자 스키마에 중복 전화번호 검증 기능 추가
- ReceiverAddModal에서 중복 전화번호 체크 로직 제거 및 메인 폼에 데이터 반영 로직 수정

SeHIgh committed:
refactor(router): 라우터 및 페이지 임포트 경로 정리

- 라우터 파일에서 페이지 컴포넌트 임포트 경로를 간소화하여 가독성 향상
- 불필요한 주석 및 코드 정리

SeHIgh commented:
꼼꼼하게 봐주셔서 정말 감사합니다!
코멘트 주신대로 Entities 계층에 data, schema 의 segment 폴더를 추가로 두어 성격별로 분리를 추가로 진행하였고,
기존에 전화번호 중복 검증을 주문하기 스키마에서 zod의 superRefine 을 이용하여 관리하도록 개선하였습니다!
(기존 refine 만으로는 여러 폼에서 정확한 위치의 중복여부를 나타낼 수 없다하여 superRefine을 이용하였습니다!)

추가로 불필요한 주석 제거 및 파일 경로 수정에 따른 import 경로 수정 작업도 진행하였습니다!

terajh reviewed (approved):
엄청 빠르게 코멘트 반영해주셨네요! 손이 정말 빠르신 것 같아요 😊
FSD 기반으로 세그먼트를 잘 나눠주신 것도 확인했고, 전화번호 중복 유효성 검증 로직을 스키마로 깔끔하게 분리해주신 점도 인상적이었습니다. PR은 승인 후 머지하도록 하겠습니다.

고생 많으셨습니다! 다음미션도 파이팅입니다~ 🙌
